libtenzir/builtins/operators/from_load_read.cpp

```diff
@@ -82,22 +85,24 @@ class read_operator final : public crtp_operator<read_operator> {
     return "read";
   }
 
-  auto optimize(expression const& filter, event_order order) const
+  auto optimize(expression const& filter, event_order order,
+                select_optimization const& selection) const
     -> optimize_result override {
     (void)filter;
-    if (order == event_order::ordered) {
+    auto parser_opt = parser_->optimize(filter, order, selection);
+    if (!parser_opt.replacement) {
```

Comment:
What happens when `parser_opt.selection_optimized` is true and `parser_opt.replacement` is false?

Reply:
This shouldn't happen because the parser (feather, parquet etc.) shouldn't know the selection information before the `optimize` function. Thus, if there is no replacement, the selection information was never used and the `parser` class is the same.

Reply:
I see. The problem is then that the return value of `parser_->optimize(...)` does not encode this invariant as part of its data structure. It is rather easy to write an optimize function for a parser with the current design that breaks your invariant. It really shouldn't be possible to do that in the first place.